# Graded Challange 3

## Introduction

Name :  Ade William Tabrani

Batch: FTDS - RMT - 026

This program is made to scrape data of Tokopedia e-commerce website regarding sales of "Seblak", an Indonesian food. The data is then cleaned/wrangled and analyzed.

Assignment objective: 

to test students' ability in webscraping specific parts of the website, to test students' ability in data wrangling, data exploration, and descriptive + inferential statistics, and to test students' analysis skill of statistics and business understanding.

## Answers

In [1]:
# set up

import pandas as pd
import scipy as sp


### Webscraping Blocks

DON'T RUN THIS.
It was specified that the Webscraping blocks need not to be run again, so I make them all into a comment, so that you can Run All and and get the data/Pandas parts without the web-scraping parts.

In [53]:
# setup
from bs4 import BeautifulSoup
from selenium import webdriver
driver = webdriver.Edge("C:/Users/Lewell/msedgedriver.exe") #change this to your 
url="https://www.tokopedia.com/search?navsource=&page=1&q=seblak&srp_component_id=02.01.00.00&srp_page_id=&srp_page_title=&st="
driver.get(url)
html = driver.page_source

soup = BeautifulSoup(html, "html.parser") #Menentukan parser HTML yang ingin kita gunakan.
print(soup.prettify()[:700])

<html data-rh="lang,translate" lang="id" translate="no">
 <head>
  <title>
   Jual seblak | Tokopedia
  </title>
  <meta charset="utf-8"/>
  <meta content="initial-scale=1, minimum-scale=1, maximum-scale=5, user-scalable=yes, width=device-width" name="viewport"/>
  <link crossorigin="" href="https://assets.tokopedia.net" rel="preconnect"/>
  <link href="https://assets.tokopedia.net" nonce="" rel="dns-prefetch"/>
  <link crossorigin="" href="https://gql.tokopedia.com" rel="preconnect"/>
  <link href="https://gql.tokopedia.com" nonce="" rel="dns-prefetch"/>
  <link crossorigin="" href="https://googleads.g.doubleclick.net" rel="preconnect"/>
  <link href="https://googleads.g.doubleclick.net" no


In [54]:
# setting up empty lists to store values from webscraping
nameProd = []
priceProd = []
sellerProd = []
townProd = []
amountProd = []
ratingProd = []

# webscraping part:
for i in range(1,21):
    #the part to continue "turning the page"
    url="https://www.tokopedia.com/search?navsource=&page={}&q=seblak&srp_component_id=02.01.00.00&srp_page_id=&srp_page_title=&st=".format(i)
    driver.get(url)
    html = driver.page_source
    soup = BeautifulSoup(html, "html.parser")

    # getting the name:
    for member in soup.find_all('div', {'class':'css-uwyh54'}):
        try:
            nameProd += [nameProd.get_text() for nameProd  in member.find('div', {'class':'prd_link-product-name css-3um8ox'})]
        except:
            nameProd.append(None)

    # getting the price:
        try:
            priceProd += [priceProd.get_text() for priceProd in member.find('div', {'class':'prd_link-product-price css-h66vau'})]
        except:
            priceProd.append(None)

    # getting the seller name:
        try:
            sellerProd += [sellerProd.get_text() for sellerProd in member.find('span', {'class':'prd_link-shop-name css-1kdc32b flip'})]
        except:
            sellerProd.append(None)

    # getting the seller town:
        try:
            townProd += [townProd.get_text() for townProd in member.find('span', {'class':'prd_link-shop-loc css-1kdc32b flip'})]
        except:
            townProd.append(None)
    
    # getting the amount sold:
        try:
            amountProd += [amountProd.get_text() for amountProd in member.find('span', {'class':'prd_label-integrity css-1sgek4h'})]
        except:
            amountProd.append(None)

    # gettting rating:
        try:
            ratingProd += [ratingProd.get_text() for ratingProd in member.find('span', {'class':'prd_rating-average-text css-t70v7i'})]
        except:
            ratingProd.append(None)
    


# # checking if length all same so turning to pandas dataframe won't be a problem
# print('amount of data in column nameProd: ', len(nameProd))
# print('amount of data in column priceProd: ', len(priceProd))
# print('amount of data in column sellerProd: ', len(sellerProd))
# print('amount of data in column townProd: ', len(townProd))
# print('amount of data in column amountProd: ', len(amountProd))
# print('amount of data in column ratingProd: ', len(ratingProd))

In [55]:
# #setting a DataFrame and populating the columns then seeing the data

seblakDF = pd.DataFrame()

seblakDF['Name'] = nameProd
seblakDF['Price'] = priceProd
seblakDF['Seller'] = sellerProd
seblakDF['Amount Sold'] = amountProd
seblakDF["Seller's town"] = townProd
seblakDF['Rating'] = ratingProd

seblakDF

Name     Price  \
0    Geli Food Seblak Ceker Tanpa Tulang Super Peda...  Rp22.500   
1                                  Kylafood Seblak Cup  Rp11.000   
2              Seblak Rafael, Seblak Coet Instan Halal  Rp25.000   
3                   KERUPUK KRUPUK SEBLAK PEDAS 250 GR  Rp17.248   
4    KRUPUK STIK RASA UDANG CAP PADI KAPAS KHAS IND...  Rp19.050   
..                                                 ...       ...   
695  Gelifood Seblak Instan Kerupuk Mawar Bumbu Ken...  Rp13.000   
696                      Seblak Instan Pedas Home Made   Rp3.500   
697           Seblak Instan Komplit Pedas Gurih Nikmat   Rp5.999   
698  seblak kering pedas 250 gr/seblak kencur/kerup...   Rp9.999   
699  seblak ceker 250 gr pedas daun jeruk, gurih,ce...   Rp9.999   

                Seller   Amount Sold      Seller's town Rating  
0             Lidigeli    29 terjual         Kab. Garut    5.0  
1     Kylafood Jakarta  100+ terjual    Jakarta Selatan    4.9  
2    Brother Meat Shop   40+ terjual              Depok    5.0  
3      TOKO KUE MEISYA  100+ terjual      Jakarta Pusat    4.9  
4             dariimyu  100+ terjual     Kab. Indramayu    5.0  
..                 ...           ...                ...    ...  
695           Lidigeli  100+ terjual         Kab. Garut    4.9  
696         the Dhecip  2rb+ terjual  Tangerang Selatan    4.8  
697     Baso Aci Ayang  2rb+ terjual         Kab. Garut    4.9  
698            aghniya  250+ terjual  Tangerang Selatan    4.8  
699            aghniya   60+ terjual  Tangerang Selatan    4.8  

[700 rows x 6 columns]

In [40]:
# turning to csv
seblakDF.to_csv("C:/Users/Lewell/Documents/GitHub/AdeWT-Hacktiv8-things/seblak.csv") 
# I have uploaded this to my public repo and will load csv to dataframe from the raw link so that y'all can run them from anywhere

Data Wrangling, Data Exploration, Descriptive and Inferential Statistics, and Analysis

#### 1. Do some simple data exploration, and tell us what insight you got:

How many rows? Insight?
Summary of data? Insight?
Check for missing value. Insight?

In [40]:
# reading the data from upload csv

SebDF = pd.read_csv('https://raw.githubusercontent.com/AdeWT/AdeWT-Hacktiv8-things/main/seblak.csv', delimiter=',')
SebDF.head()

Unnamed: 0                                               Name     Price  \
0           0  Gelifood Seblak Instan Kerupuk Mawar Bumbu Ken...  Rp13.000   
1           1                                Kylafood Seblak Cup  Rp11.000   
2           2                                      Seblak Instan   Rp8.000   
3           3                              Seblak Sultan instant  Rp14.000   
4           4  KRUPUK STIK RASA UDANG CAP PADI KAPAS KHAS IND...  Rp19.050   

             Seller   Amount Sold    Seller's town  Rating  
0          Lidigeli  100+ terjual       Kab. Garut     4.9  
1  Kylafood Jakarta  100+ terjual  Jakarta Selatan     4.9  
2     ROEMAH MOSLEM    19 terjual    Kab. Boyolali     5.0  
3  Sultan Meat Shop  250+ terjual            Medan     4.8  
4          dariimyu  100+ terjual   Kab. Indramayu     5.0

In [41]:
# getting summary:

SebDF.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 350 entries, 0 to 349
Data columns (total 7 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   Unnamed: 0     350 non-null    int64  
 1   Name           350 non-null    object 
 2   Price          350 non-null    object 
 3   Seller         349 non-null    object 
 4   Amount Sold    344 non-null    object 
 5   Seller's town  349 non-null    object 
 6   Rating         337 non-null    float64
dtypes: float64(1), int64(1), object(5)
memory usage: 19.3+ KB


Insight:

There are total 350 rows of data, with most of the 6 columns shown to be having null values. They may need to be removed for statistics and when looking deeper into the specifics. Data types are off, with the only correct one being Rating of float64. 

#### 2. Data cleaning: 

Handle missing values, get the data types correct, and any other data cleaning you see fit.

As the data will be used for exploring ratings, amount of sales, and seller's town, it's irrelevant to keep empty data, hence the rows where those columns are empty will be purged from the table.

In [42]:
# dropping missing values

SebDF.dropna(inplace=True)

#to see amount of rows and missing values remaining:
SebDF.info()

<class 'pandas.core.frame.DataFrame'>
Index: 336 entries, 0 to 349
Data columns (total 7 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   Unnamed: 0     336 non-null    int64  
 1   Name           336 non-null    object 
 2   Price          336 non-null    object 
 3   Seller         336 non-null    object 
 4   Amount Sold    336 non-null    object 
 5   Seller's town  336 non-null    object 
 6   Rating         336 non-null    float64
dtypes: float64(1), int64(1), object(5)
memory usage: 21.0+ KB


Insight: 14 rows with missing values have been removed, data now consists only of data with zero null.

We can drop the index/unnamed column as it's not going to be used for our exploration / analysis.

In [43]:
# dropping the column unnamed

SebDF.drop('Unnamed: 0', axis='columns', inplace=True)

#to see the change
SebDF.info()

<class 'pandas.core.frame.DataFrame'>
Index: 336 entries, 0 to 349
Data columns (total 6 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   Name           336 non-null    object 
 1   Price          336 non-null    object 
 2   Seller         336 non-null    object 
 3   Amount Sold    336 non-null    object 
 4   Seller's town  336 non-null    object 
 5   Rating         336 non-null    float64
dtypes: float64(1), object(5)
memory usage: 18.4+ KB


Let's clear out the formatting of the numerical data so that it can be turned into int/float after that.

In [44]:
# check how data is formatted:

SebDF.iloc[1:10]

Name     Price  \
1                                 Kylafood Seblak Cup  Rp11.000   
2                                       Seblak Instan   Rp8.000   
3                               Seblak Sultan instant  Rp14.000   
4   KRUPUK STIK RASA UDANG CAP PADI KAPAS KHAS IND...  Rp19.050   
5                       Seblak Instan Pedas Home Made   Rp3.500   
6                                 Kylafood Seblak Cup  Rp11.000   
8   Seblak Instan Ceu Nthien Khas Bandung Rasana N...  Rp15.000   
9                                       seblak frozen  Rp10.400   
10                      Seblak Instan Pedas Home Made   Rp3.500   

                      Seller   Amount Sold      Seller's town  Rating  
1           Kylafood Jakarta  100+ terjual    Jakarta Selatan     4.9  
2              ROEMAH MOSLEM    19 terjual      Kab. Boyolali     5.0  
3           Sultan Meat Shop  250+ terjual              Medan     4.8  
4                   dariimyu  100+ terjual     Kab. Indramayu     5.0  
5                 the Dhecip  2rb+ terjual  Tangerang Selatan     4.8  
6           Kylafood Jakarta  100+ terjual    Jakarta Selatan     4.9  
8   Central Seblak Nusantara  2rb+ terjual  Tangerang Selatan     4.9  
9        pelangi girls naget  2rb+ terjual      Jakarta Timur     4.9  
10                the Dhecip  2rb+ terjual  Tangerang Selatan     4.8

Insight :

- Price: remove 'Rp' and dot
- Amount sold: remove '+' and the word 'terjual', turn 'rb' into 000

In [45]:
#cleaning price column
SebDF['Price'] = SebDF['Price'].str.replace('Rp','')
SebDF['Price'] = SebDF['Price'].str.replace('.','')

#cleaning amount sold column
SebDF['Amount Sold'] = SebDF['Amount Sold'].str.replace('+','')
SebDF['Amount Sold'] = SebDF['Amount Sold'].str.replace(' terjual','')
SebDF['Amount Sold'] = SebDF['Amount Sold'].str.replace('rb', '000')

#to check the change
SebDF.iloc[1:10]

Name  Price  \
1                                 Kylafood Seblak Cup  11000   
2                                       Seblak Instan   8000   
3                               Seblak Sultan instant  14000   
4   KRUPUK STIK RASA UDANG CAP PADI KAPAS KHAS IND...  19050   
5                       Seblak Instan Pedas Home Made   3500   
6                                 Kylafood Seblak Cup  11000   
8   Seblak Instan Ceu Nthien Khas Bandung Rasana N...  15000   
9                                       seblak frozen  10400   
10                      Seblak Instan Pedas Home Made   3500   

                      Seller Amount Sold      Seller's town  Rating  
1           Kylafood Jakarta         100    Jakarta Selatan     4.9  
2              ROEMAH MOSLEM          19      Kab. Boyolali     5.0  
3           Sultan Meat Shop         250              Medan     4.8  
4                   dariimyu         100     Kab. Indramayu     5.0  
5                 the Dhecip        2000  Tangerang Selatan     4.8  
6           Kylafood Jakarta         100    Jakarta Selatan     4.9  
8   Central Seblak Nusantara        2000  Tangerang Selatan     4.9  
9        pelangi girls naget        2000      Jakarta Timur     4.9  
10                the Dhecip        2000  Tangerang Selatan     4.8

The column of price and amount sold can now be turned into integer (there's no comma on prices and amount sold, so there's no need for float type).

In [46]:
#turning column price and amount into int

SebDF['Price'] = SebDF['Price'].astype(int)
SebDF['Amount Sold'] = SebDF['Amount Sold'].astype(int)

#to see the change
SebDF.info()

<class 'pandas.core.frame.DataFrame'>
Index: 336 entries, 0 to 349
Data columns (total 6 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   Name           336 non-null    object 
 1   Price          336 non-null    int32  
 2   Seller         336 non-null    object 
 3   Amount Sold    336 non-null    int32  
 4   Seller's town  336 non-null    object 
 5   Rating         336 non-null    float64
dtypes: float64(1), int32(2), object(3)
memory usage: 15.8+ KB


In [26]:
#just checking to see if all data are formatted correctly 
#by picking data from the middle

SebDF.iloc[200:216]

Name   Price  \
204  PROMO !!!! Seblak Instan Beli 20 free 1 Murah ...    4550   
205                      Seblak Instan Pedas Home Made    3500   
206  Gelifood Seblak Instan Kerupuk Mawar Bumbu Ken...   13000   
207           Seblak Instan Komplit Pedas Gurih Nikmat    5999   
208  seblak kering pedas 250 gr/seblak kencur/kerup...    9999   
209  seblak ceker 250 gr pedas daun jeruk, gurih,ce...    9999   
212  Seblak Instant Carla With Batagor dan Siomay J...  150000   
213                           Kylafood Seblak Original   21500   
214   Pasar Keripik Cimol Seblak Pedas size Kecil 130g   12500   
215  Makanan ringan pedas, Keripik kaca/Kerupuk seb...   10000   
216                                     Seblak Basreng    6000   
218        SEBLAK CEKER CAP DUA PUTRA ISI 20 X 10 gram    9000   
219  SEBLAK INSTAN ORI/SEBLAK INSTAN KERUPUK/SEBLAK...    3899   
220                      Seblak Instan Pedas Home Made    3500   
221  Gelifood Seblak Instan Kerupuk Mawar Bumbu Ken...   13000   
222                         seblak instant sajodo snak   19000   

                         Seller  Amount Sold      Seller's town  Rating  
204               ivy's-kitchen          500     Kab. Tangerang     4.7  
205                  the Dhecip         2000  Tangerang Selatan     4.8  
206                    Lidigeli          100         Kab. Garut     4.9  
207              Baso Aci Ayang         2000         Kab. Garut     4.9  
208                     aghniya          250  Tangerang Selatan     4.8  
209                     aghniya           60  Tangerang Selatan     4.8  
212                Seblak carla          750          Tangerang     5.0  
213                    kylafood        10000            Bandung     4.9  
214       Pasar Keripik Bandung          500            Bandung     4.9  
215                   sansay327           18  Tangerang Selatan     4.8  
216  Official Snack Murah Batam          100              Batam     4.6  
218                 HRNWN STORE            5     Kab. Tangerang     5.0  
219                   dakar_ofc           80     Kab. Mojokerto     4.2  
220                  the Dhecip         2000  Tangerang Selatan     4.8  
221                    Lidigeli          100         Kab. Garut     4.9  
222          hikmah berkah jaya          750      Jakarta Barat     4.8

Let's check for duplicates.

In [48]:
#to check:
truCount = 0
falCount = 0
for i in SebDF.duplicated():
    if i == True: truCount += 1
    if i == False: falCount += 1

print(f'There are {truCount} duplicates')
print(f'There are {truCount} non-duplicates')

There are 245 duplicates
There are 245 non-duplicates


Since we don't need duplicates because 1, they tell the same story, 2, they can sway statistics, let's drop the duplicated data.

In [50]:
SebDF.drop_duplicates(inplace=True)

In [51]:
SebDF.info()

<class 'pandas.core.frame.DataFrame'>
Index: 91 entries, 0 to 324
Data columns (total 6 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   Name           91 non-null     object 
 1   Price          91 non-null     int32  
 2   Seller         91 non-null     object 
 3   Amount Sold    91 non-null     int32  
 4   Seller's town  91 non-null     object 
 5   Rating         91 non-null     float64
dtypes: float64(1), int32(2), object(3)
memory usage: 4.3+ KB


#### 3. Problem Statement

So, I'm a Hacktiv8 FTDS alumni who is looking to sell seblak to increase my income. I would love to increase my revenue up to 2 million IDR per month from selling seblak by the time I'm 6 months into the business, the increased income would affect me, my family, and my friends as more money means I can go on more trips/hangouts with them, pay people to do chores and other stuff, and I would be able to pamper my friends and family too. 

I have scraped (aka doing illegal thing) the Tokopedia e-commerce website to see if the seblak trend is just another fad or perhaps it has big enough demand to support my goal. I have also cleaned the data with pandas in python.

Let's look at the descriptives first.

#### 4. Analysis

##### 1. Count mean, median, standard deviation, skewness, curtosis from Price column, Amount Sold column, and Rating. What insight can you gain from such statistics? Note: you don't have to handle outliers.

##### 2. Think about how much can you get from selling seblak. Think about the floor and ceiling of revenue if you're selling seblak. Use confidence interval to gain lower value and upper value from the population distribution of revenue.

Assume data is distributed normally and that the sales are of per month data.

##### 3. Are prices outside and inside of Jabodetabek different, considering the difference of material price between two locations?

Use hypothesis testing (h0 = ... , h1 = ...) and write your own hypotheses.

##### 4. Do people like prices that are cheaper?

You can use h0 and h1 too, with comparison of prices.